Improve infrastructure repositories documentation (#2119)
conda-forge/conda-forge.github.io: docs/maintainer/infrastructure.md

- ⚙️ Deployed in [Github Actions via `regro/cf-graph-countyfair`](https://github.com/regro/cf-graph-countyfair)
- ⛓ Needs [`regro/cf-scripts`](#regrocf-scripts), [`conda-forge/conda-forge-pinning-feedstock`](#conda-forge-pinning)
- 🤖 Uses [`@regro-cf-autotick-bot`](#bot-accounts)
- 🔒 Has access to Github API

The logic to build the graph is provided by [`cf-scripts`](#regrocf-scripts).

#### docker-images

This repository builds the Docker images used to provide a unified system on all Linux builds.

- ⚙️ Deployed in [`conda-forge/docker-images`](https://github.com/conda-forge/docker-images)
- 🔒 Has access to [DockerHub](#docker-hub) and [Quay.io](#quay)
- ⛓ Needed by `staged-recipes`, feedstocks.

### Code repositories

These are repositories that hold programs and other codes that define behavior.
However, their actions are often not triggered here, but rather used by other parts of the conda-forge ecosystem.

#### Smithy

This is the main feedstock creation and maintenance tool.

- 📜 Source at [`conda-forge/conda-smithy`](https://github.com/conda-forge/conda-smithy)
- 📦 Packaged at [`conda-forge/conda-smithy-feedstock`](https://github.com/conda-forge/conda-smithy-feedstock)

> - 📖 [Documentation](https://github.com/conda-forge/conda-smithy/blob/main/README.md)

Reviewer: ```suggestion
- 📖 [Documentation](https://github.com/conda-forge/conda-smithy/blob/main/README.md)
- 📖 [conda-forge.yml Documentation](https://conda-forge.org/docs/maintainer/conda_forge_yml/)
```

title: 'Infrastructure'
---

# Infrastructure

> 

Author: It is :)

#### Workflows

The main job run on `staged-recipes` is the `conda-build` job that runs on every PR (and push to `main`) to check whether the recipes build packages correctly.
These jobs run on Azure Pipelines defined in [`.azure-pipelines/`](https://github.com/conda-forge/staged-recipes/tree/main/.azure-pipelines).

The actual creation of the feedstock is run in [conda-forge/admin-requests](#admin-requests).

Additional workflows help users set up their recipes correctly. They react to events in PRs:

- [`automate-review-labels`](https://github.com/conda-forge/staged-recipes/blob/main/.github/workflows/automate-review-labels.yml): Updates PR labels to streamline reviews and requests for help.
- [`correct_directory`](https://github.com/conda-forge/staged-recipes/blob/main/.github/workflows/correct_directory.yml): Posts a PR comment if `meta.yaml` and friends were not added in a `recipes/` subdirectory.
- [`do_not_edit_example`](https://github.com/conda-forge/staged-recipes/blob/main/.github/workflows/do_not_edit_example.yml): Posts a PR comment if the `recipes/example/` recipe was edited.

External services connect to `staged-recipes` too:

> - The `@conda-forge-linter` bot (deployed at [`webservices`](#webservices)) will lint and provide hints in PRs based on the contents of the recipe.

Reviewer: ```suggestion
- The `@conda-forge-admin` bot (deployed at [`webservices`](#webservices)) will lint and provide hints in PRs based on the contents of the recipe.
```

- 🔒 Has access to Azure, Anaconda.org (cf-staging)

#### conda-forge-ci-setup

This special feedstock provides a package that defines the logic to install and configure a common CI setup across providers.

- ⚙️ Deployed in [Anaconda.org](https://anaconda.org/conda-forge/conda-forge-ci-setup) via [`conda-forge/conda-forge-ci-setup-feedstock`](https://github.com/conda-forge/conda-forge-ci-setup-feedstock)
- 🔒 Has access to Azure, Anaconda.org (cf-staging)

#### regro/cf-graph-countyfair

This is the graph data used by [`autotick-bot`](#autotick-bot).

- ⚙️ Deployed in [Github Actions via `regro/cf-graph-countyfair`](https://github.com/regro/cf-graph-countyfair)
- ⛓ Needs [`regro/cf-scripts`](#regrocf-scripts), [`conda-forge/conda-forge-pinning-feedstock`](#conda-forge-pinning)
- 🤖 Uses [`@regro-cf-autotick-bot`](#bot-accounts)
- 🔒 Has access to Github API

The logic to build the graph is provided by [`cf-scripts`](#regrocf-scripts).

#### regro/conda-suggest-conda-forge

- [`regro/conda-suggest-conda-forge`](https://github.com/regro/conda-suggest-conda-forge) provides [`conda-suggest`](https://github.com/conda-incubator/conda-suggest) files that map executables to package names.

> 

Reviewer: As far as I know, this is a private-ish tool built by @scopatz and not so much a semi-official conda-forge thing. It might make sense to not list it here.

#### docker-images

This repository builds the Docker images used to provide a unified system on all Linux builds.

- ⚙️ Deployed in [`conda-forge/docker-images`](https://github.com/conda-forge/docker-images)
- 🔒 Has access to [DockerHub](#docker-hub) and [Quay.io](#quay)
- ⛓ Needed by `staged-recipes`, feedstocks.

### Code repositories

These are repositories that hold programs and other codes that define behavior.
However, their actions are often not triggered here, but rather used by other parts of the conda-forge ecosystem.

#### Smithy

This is the main feedstock creation and maintenance tool.

- 📜 Source at [`conda-forge/conda-smithy`](https://github.com/conda-forge/conda-smithy)
- 📦 Packaged at [`conda-forge/conda-smithy-feedstock`](https://github.com/conda-forge/conda-smithy-feedstock)

> - 📖 [Documentation](https://github.com/conda-forge/conda-smithy/blob/main/README.md)

Reviewer: We should mention the conda-forge docs as well since that is where all of the smithy options in the conda-forge.yml are exposed in a website.

Most of its usage is automated by our infrastructure:

- Feedstock creation and services registration at [`staged-recipes`](#staged-recipes)

> - Regeneration (rerendering), linting and hinting in PRs done by `conda-forge-linter` on [`webservices`](#web-services)

Reviewer: ```suggestion
- Regeneration (rerendering), linting and hinting in PRs done by `conda-forge-admin` on [`webservices`](#web-services)
```

title: 'Infrastructure'
---

# Infrastructure

## Repositories

### Staged-recipes

This repository is the gateway to conda-forge and where users can submit new recipes which, once reviewed and accepted, will generate a new feedstock and team.
You can find the detailed guide for submitting new package recipes in [The staging process](adding_pkgs.md#creating-recipes).

- ⚙️ Deployed in [`conda-forge/staged-recipes`](https://github.com/conda-forge/staged-recipes)

> - 🔒 Has access to Azure Pipelines, Github Actions, Travis CI, Anaconda.org (cf-staging)

Reviewer: ```suggestion
- 🔒 Has access to Anaconda.org (cf-staging)
```

- `compute_build_graph.py`: Supports `build_all.py` by providing the job graph with all the submitted recipes.

The YAML files included in `.ci_support` are minimal and not rendered like the ones you find in feedstocks.
Instead, conda-build will take these and combine them with the pinnings from `conda-forge-pinning` at runtime.
Also note that `staged-recipes` only builds for x64. Support for additional architectures can only be done once a feedstock has been provided.

- Linux: `linux64.yaml` plus the CUDA (10.2, 11.0, 11.1 and 11.2) variants.
- macOS: `osx64.yaml`.
- Windows `win64.yaml`.

The directory `.scripts` contains roughly the same shell scripts that would be used in a feedstock for the CI pipelines.
However, since `staged-recipes` does not support rerendering, these are kept in sync manually and it is common to see some differences.

#### Workflows

Two main jobs run on `staged-recipes``:

- The `conda-build` jobs that run on every PR (and push to `main`) check whether the recipes build packages correctly.
  These jobs run on Azure Pipelines defined in [`.azure-pipelines/`](https://github.com/conda-forge/staged-recipes/tree/main/.azure-pipelines).

> - The [`create_feedstocks` workflow](https://github.com/conda-forge/staged-recipes/blob/main/.github/workflows/create_feedstocks.yml) runs after each push to `main` (and every 10 minutes) to create the new feedstock repositories on the `conda-forge` organization.

Reviewer: This runs in admin requests

Please open a [PR](../glossary.md#term-PR) and/or an issue there, if you think a pin needs to be advanced. For more information on updating globally pinned packages, please refer to [Updating package pins](pinning_deps.md#update-pins).

### conda-forge-repodata-patches

This repository creates the `repodata.json` patches used by the Anaconda.org to amend the metadata coming from the published packages.

- ⚙️ Deployed in [Anaconda.org](https://anaconda.org/conda-forge/conda-forge-repodata-patches) via [`conda-forge/conda-forge-repodata-patches-feedstock`](https://github.com/conda-forge/conda-forge-repodata-patches-feedstock)
- 🔒 Has access to Azure, Anaconda.org (cf-staging)

### conda-forge-ci-setup

This special feedstock provides a package that defines the logic to install and configure a common CI setup across providers.

- ⚙️ Deployed in [Anaconda.org](https://anaconda.org/conda-forge/conda-forge-ci-setup) via [`conda-forge/conda-forge-ci-setup-feedstock`](https://github.com/conda-forge/conda-forge-ci-setup-feedstock)
- 🔒 Has access to Azure, Anaconda.org (cf-staging)

> ### Documentation

Reviewer 2: Should we rename this to "Website" and leave an anchor with `#documentation` right on top?

Author: Now that things are consolidating, I want to move the repos listed below under "Automatic maintenance" up here, reorder and categorize all the repos so that the more important ones (staged-recipes, documentation) are those most easily found by users. That may bring some changes in headings; let's defer discussions about final names until then, ok?

Author: Done. I did rename this to Website, so I'm resolving this conversation. Please comment again if you would like further changes.

The current [conda-forge.org](https://conda-forge.org) is a statically generated website published to Github Pages.

- 📜 Source at [conda-forge/conda-forge.github.io](https://github.com/conda-forge/conda-forge.github.io/)
- ⚙️ Deployed in [conda-forge.org](https://conda-forge.org/)

> 

Reviewer 2: We probably need to mention these three services too:

- PR Previews at [Netlify](https://app.netlify.com/sites/conda-forge-previews)
- Statistics at [GoatCounter](https://conda-forge.goatcounter.com/)
- Search backend at [Algolia](https://dashboard.algolia.com/apps/KB43FQOB7U/dashboard)
- Status JSONS at `conda-forge/status`
- Packages-to-feedstocks in `conda-forge/feedstock-outputs`

The core logic is defined in the Python script [`.github/workflows/scripts/create_feedstocks.py`](https://github.com/conda-forge/staged-recipes/blob/main/.github/workflows/scripts/create_feedstocks.py).

#### autotick-bot

- ⚙️ Deployed in [`regro/cf-scripts`](https://github.com/regro/cf-scripts)
- ⛓ Needs [`regro/cf-scripts`](https://github.com/regro/cf-scripts), [`regro/cf-graph-countyfair`](https://github.com/regro/cf-graph-countyfair), [`conda-forge/conda-forge-pinning-feedstock`](https://github.com/conda-forge/conda-forge-pinning-feedstock)
- 🤖 Uses [`@regro-cf-autotick-bot`](https://github.com/regro-cf-autotick-bot)

:::note
The older repo [`regro/autotick-bot`](https://github.com/regro/autotick-bot) is no longer in use;
the bot now runs directly in `regro/cf-scripts`.
:::

#### webservices

- ⚙️ Deployed in Heroku Dyno (`conda-forge.herokuapp.com`)
- ⛓ Needs [`conda-forge/conda-forge-webservices`](https://github.com/conda-forge/conda-forge-webservices)

> - 🤖 Uses [`@conda-forge-webservices`](https://github.com/apps/conda-forge-webservices), [`@conda-forge-admin`](https://github.com/conda-forge-admin), [`@conda-forge-linter`](https://github.com/conda-forge-linter)

Reviewer: ```suggestion
- 🤖 Uses [`@conda-forge-webservices`](https://github.com/apps/conda-forge-webservices), [`@conda-forge-admin`](https://github.com/conda-forge-admin)
```

#### admin-migrations

- ⚙️ Deployed in [Github Actions via `conda-forge/admin-migrations`](https://github.com/conda-forge/admin-migrations)
- 🤖 Uses [`@conda-forge-curator`](https://github.com/apps/conda-forge-curator)
- 🔒 Has access to Github API, Anaconda.org (conda-forge and cf-staging), Circle, Travis, Azure

This repository hosts workflows that are running 24/7.
Its job is to procure an automation loop where some maintenance tasks are added.
Its main user is the core team.

#### admin-requests

- ⚙️ Deployed in [Github Actions via `conda-forge/admin-requests`](https://github.com/conda-forge/admin-requests)
- 🤖 Uses [`@conda-forge-curator`](https://github.com/apps/conda-forge-curator)
- 🔒 Has access to Github API, Anaconda.org (conda-forge and cf-staging), Circle, Travis, Azure

This repository hosts workflows that mainly run when triggered by a user-initiated action.
This is usually done via a PR that, once approved, is merged and triggers the requested action (mark a package as broken, archive a feedstock, etc).

It also does the job of creating new feedstocks for recipes that have been merged in [`conda-forge/staged-recipes`](#staged-recipes).

> The [`create_feedstocks` workflow](https://github.com/conda-forge/admin-requests/blob/main/.github/workflows/create_feedstocks.yml) runs every 10 minutes to create the new feedstock repositories on the `conda-forge` organization.

Reviewer: ```suggestion
The [`create_feedstocks` workflow](https://github.com/conda-forge/admin-requests/blob/main/.github/workflows/create_feedstocks.yml) runs several times per hour to create the new feedstock repositories on the `conda-forge` organization.
```

The core logic is defined in the Python script [`.github/workflows/scripts/create_feedstocks.py`](https://github.com/conda-forge/staged-recipes/blob/main/.github/workflows/scripts/create_feedstocks.py).

#### autotick-bot

- ⚙️ Deployed in [`regro/cf-scripts`](https://github.com/regro/cf-scripts)
- ⛓ Needs [`regro/cf-scripts`](https://github.com/regro/cf-scripts), [`regro/cf-graph-countyfair`](https://github.com/regro/cf-graph-countyfair), [`conda-forge/conda-forge-pinning-feedstock`](https://github.com/conda-forge/conda-forge-pinning-feedstock)
- 🤖 Uses [`@regro-cf-autotick-bot`](https://github.com/regro-cf-autotick-bot)

:::note
The older repo [`regro/autotick-bot`](https://github.com/regro/autotick-bot) is no longer in use;
the bot now runs directly in `regro/cf-scripts`.
:::

#### webservices

- ⚙️ Deployed in Heroku Dyno (`conda-forge.herokuapp.com`)
- ⛓ Needs [`conda-forge/conda-forge-webservices`](https://github.com/conda-forge/conda-forge-webservices)

> - 🤖 Uses [`@conda-forge-webservices`](https://github.com/apps/conda-forge-webservices), [`@conda-forge-admin`](https://github.com/conda-forge-admin), [`@conda-forge-linter`](https://github.com/conda-forge-linter)

Reviewer: We don't use tokens from conda-forge-linter anymore as far as I know.

Reviewer: We also have a webservices GitHub app that does most GitHub API related tasks now.

- [`help-c-cpp`](https://github.com/orgs/conda-forge/teams/help-c-cpp)
- [`help-cdts`](https://github.com/orgs/conda-forge/teams/help-cdts)
- [`help-go`](https://github.com/orgs/conda-forge/teams/help-go)
- [`help-java`](https://github.com/orgs/conda-forge/teams/help-java)
- [`help-julia`](https://github.com/orgs/conda-forge/teams/help-julia)
- [`help-nodejs`](https://github.com/orgs/conda-forge/teams/help-nodejs)
- [`help-osx-arm64`](https://github.com/orgs/conda-forge/teams/help-osx-arm64)
- [`help-perl`](https://github.com/orgs/conda-forge/teams/help-perl)
- [`help-ppc64le`](https://github.com/orgs/conda-forge/teams/help-ppc64le)
- [`help-pypy`](https://github.com/orgs/conda-forge/teams/help-pypy)
- [`help-python`](https://github.com/orgs/conda-forge/teams/help-python)
- [`help-python-c`](https://github.com/orgs/conda-forge/teams/help-python-c)
- [`help-r`](https://github.com/orgs/conda-forge/teams/help-r)
- [`help-ruby`](https://github.com/orgs/conda-forge/teams/help-ruby)
- [`miniforge`](https://github.com/orgs/conda-forge/teams/miniforge)
- [`all-members`](https://github.com/orgs/conda-forge/teams/all-members)

#### Configuration

> - [`conda-forge/.github`](https://github.com/conda-forge/.github): Organization-wide configuration, profile information, etc.

Reviewer 2: We should probably add `conda-forge/.cirun` here too.

### Continuous integration

:::tip See also
Refer to the [`conda-forge.yml` documentation](/docs/maintainer/conda_forge_yml/) to learn how to configure your CI providers.
:::

#### Azure Pipelines

- 🌐 https://dev.azure.com/conda-forge/feedstock-builds/_build
- 📍 Available on all feedstocks
- 🛠 Provides [Microsoft-hosted runners](https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml) (x64 Linux, macOS and Windows)
- 🔒 Needs access to Anaconda.org (cf-staging)

conda-forge benefits from the generously offered Microsoft-hosted runners.

#### Travis CI

- 🌐 https://app.travis-ci.com/github/conda-forge
- 📍 Available on all feedstocks
- 🛠 Provides [native Linux aarch64 and ppc64le runners](https://docs.travis-ci.com/user/reference/overview/)
- 🔒 Needs access to Anaconda.org (cf-staging)

#### Cirun

- 🌐 https://cirun.io
- 📍 Available on selected feedstocks only
- 🛠 Provides several architectures (depending on feedstock configuration)
- 🔒 Needs access to Anaconda.org (cf-staging) and the configured backend

Configured with `@conda-forge-daemon`.

> Organization-wide configuration can be found in the [`.cirun` repository](https://github.com/conda-forge/.cirun).

Reviewer 2: An admonition here about the GPU Server pointing to `Quansight/open-gpu-server` would be nice too.

- 🛠 Provides [GitHub-hosted runners](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners) (x64 Linux, macOS and Windows)
- 🔒 Has access to Github API

#### Retired services

- [AppVeyor](https://ci.appveyor.com/account/conda-forge/projects)
- Circle CI
- Drone.io

### Delivery and distribution

#### Anaconda.org

- 🌐 Channels/organizations: [`cf-staging`](https://anaconda.org/cf-staging/dashboard), [`conda-forge`](https://anaconda.org/conda-forge/dashboard)
- ⛓ Used by feedstocks

#### Docker Hub

- 🌐 https://hub.docker.com/u/condaforge/
- ⛓ Used by [`conda-forge/docker-images`](#docker-images), [`conda-forge/miniforge-images`](https://github.com/conda-forge/miniforge-images)

#### Github Packages

- 🌐 https://github.com/orgs/channel-mirrors/packages
- ⛓ Used by [`channel-mirrors/conda-oci-mirror`](https://github.com/channel-mirrors/conda-oci-mirror)

#### Github Releases

- 🌐 https://github.com/conda-forge/miniforge/releases
- ⛓ Used by [`conda-forge/miniforge`](https://github.com/conda-forge/miniforge)

#### Quay

- 🌐 https://quay.io/organization/condaforge
- ⛓ Used by [`conda-forge/docker-images`](#docker-images)

### Servers

#### Heroku

> - 🌐 https://hub.docker.com/u/condaforge/

Reviewer 2: I think this URL Is wrong.

Author: Do you know the right one, by chance?

Author: To be precise: I tried https://conda-forge.herokuapp.com but only  got a 404, though that maybe because I didn't use a proper endpoint?

Reviewer: That url is correct but we don't serve an endpoint there. See for example this one:

https://conda-forge.herokuapp.com/status-monitor

To be honest, that app is a service backend and not meant to be called directly by a human.

Author: Is there something like an admin interface that would be more meaningful to link to here? Even though that wouldn't be accessible by users, of course, I have got the impression that core members reference this page as much as any one else...

Reviewer: This is the url the status page uses: https://conda-forge.herokuapp.com/alive

Reviewer: The only one that displays data is the status-monitor one IIRC. So let's link that!

uploaded from the new feedstock, by design.
Once this is done correctly and the package is uploaded,
you can then request the conda-forge core devs to archive the old feedstock.

## Automated maintenance

These components perform actions in automated ways, either triggered by a specific event or continuously as part of a loop.

### admin-migrations

- ⚙️ Deployed in [Github Actions via `conda-forge/admin-migrations`](https://github.com/conda-forge/admin-migrations)
- 🤖 Uses [`@conda-forge-curator`](https://github.com/apps/conda-forge-curator)
- 🔒 Has access to Github API, Anaconda.org (conda-forge and cf-staging), Circle, Travis, Azure, Drone

This repository hosts workflows that are running 24/7.
Its job is to procure an automation loop where some maintenance tasks are added.
Its main user is the core team.

### admin-requests

- ⚙️ Deployed in [Github Actions via `conda-forge/admin-requests`](https://github.com/conda-forge/admin-requests)
- 🤖 Uses [`@conda-forge-curator`](https://github.com/apps/conda-forge-curator)

> - 🔒 Has access to Github API, Anaconda.org

Reviewer: This one also has access to the ci services.

This repository hosts workflows that mainly run when triggered by a user-initiated action.
This is usually done via a PR that, once approved, is merged and triggers the requested action (mark a package as broken, archive a feedstock, etc).

### artifact-validation

- ⚙️ Deployed in [Github Actions via `conda-forge/artifact-validation`](https://github.com/conda-forge/artifact-validation)
- ⛓ Needs [`regro/libcfgraph`](https://github.com/regro/libcfgraph)
- 🤖 Uses [`@conda-forge-curator`](https://github.com/apps/conda-forge-curator)
- 🔒 Has access to Github API, Anaconda.org API

The workflows (and code) to scan artifacts uploaded to [anaconda.org/conda-forge](https://anaconda.org/conda-forge).

> Its main output comes in the form of [issues in `conda-forge/artifact-validation`](https://github.com/conda-forge/artifact-validation/issues)

Reviewer: This one is deprecated too and can be removed.

### autotick-bot

> - ⚙️ Deployed in [Github Actions via `regro/autotick-bot`](https://github.com/regro/autotick-bot)

Reviewer: This repo is not used anymore. Bot jobs run right in cf-scripts.

- ⛓ Needs [`regro/cf-scripts`](https://github.com/regro/cf-scripts), [`regro/cf-graph-countyfair`](https://github.com/regro/cf-graph-countyfair), [`conda-forge/conda-forge-pinning-feedstock`](https://github.com/conda-forge/conda-forge-pinning-feedstock)
- 🤖 Uses [`@regro-cf-autotick-bot`](https://github.com/regro-cf-autotick-bot)

> - 🔒 Has access to Github API

Reviewer: The bot does not have any enhanced access to conda-forge beyond a normal user.

- [`help-python-c`](https://github.com/orgs/conda-forge/teams/help-python-c)
- [`help-r`](https://github.com/orgs/conda-forge/teams/help-r)
- [`help-ruby`](https://github.com/orgs/conda-forge/teams/help-ruby)
- [`help-rust`](https://github.com/orgs/conda-forge/teams/help-rust)
- [`miniforge`](https://github.com/orgs/conda-forge/teams/miniforge)
- [`all-members`](https://github.com/orgs/conda-forge/teams/all-members)

#### Configuration

- [`conda-forge/.github`](https://github.com/conda-forge/.github): Organization-wide configuration, profile information, etc.
- [`conda-forge/.cirun`](https://github.com/conda-forge/.cirun): Organization-wide configuration for special ci runners (e.g. GPU), c.f. [Cirun](#cirun).

#### Bot accounts

- [`conda-forge-admin`](https://github.com/conda-forge-admin)
- [`conda-forge-bot`](https://github.com/conda-forge-bot)
- [`conda-forge-coordinator`](https://github.com/conda-forge-coordinator)
- [`conda-forge-daemon`](https://github.com/conda-forge-daemon)
- [`conda-forge-linter`](https://github.com/conda-forge-linter)
- [`conda-forge-manager`](https://github.com/conda-forge-manager)
- [`conda-forge-status`](https://github.com/conda-forge-status)

> - [`regro-cf-autotick-bot`](https://github.com/regro-cf-autotick-bot)

Reviewer: -bot, -coordinator, -manager, and -status are all no longer used

I need to check on -linter

:::info
These accounts exist but are not in active usage anymore:

- [`conda-forge-drone-ci`](https://github.com/conda-forge-drone-ci)

:::

#### Apps

- `conda-forge-curator`
- `conda-forge-webservices`

:::info
These apps exist but are not in active usage anymore:

- `conda-forge drone instance`

:::

#### Workflows

> - [`beckermr/turnstyle-python`](https://github.com/beckermr/turnstyle-python): Prevents multiple CI jobs from running in parallel to avoid race conditions.

Reviewer: It might be best to transfer this to conda-forge.

- [`conda-forge/automerge-action`](https://github.com/conda-forge/automerge-action)
- [`conda-forge/github-app-token`](https://github.com/conda-forge/github-app-token) (fork)
- [`conda-forge/webservices-dispatch-action`](https://github.com/conda-forge/webservices-dispatch-action)

### Continuous integration

:::tip See also
Refer to the [`conda-forge.yml` documentation](/docs/maintainer/conda_forge_yml/) to learn how to configure your CI providers.
:::

#### Azure Pipelines

- 🌐 https://dev.azure.com/conda-forge/feedstock-builds/_build
- 📍 Available on all feedstocks
- 🛠 Provides [Microsoft-hosted runners](https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml) (x64 Linux, macOS and Windows)
- 🔒 Needs access to Anaconda.org (cf-staging)

conda-forge benefits from the generously offered Microsoft-hosted runners.

#### Travis CI

- 🌐 https://app.travis-ci.com/github/conda-forge

> - 📍 Available on all feedstocks

Reviewer: I think we have or will soon move this to be opt-in.

uploaded from the new feedstock, by design.
Once this is done correctly and the package is uploaded,
you can then request the conda-forge core devs to archive the old feedstock.

## Services & providers

### Github resources

In addition to the thousands of repositories, conda-forge uses several other Github services.

#### Organizations

- [`conda-forge`](https://github.com/conda-forge): the main organization
- [`regro`](https://github.com/regro): hosts the `autotick-bot` machinery
- [`channel-mirrors`](https://github.com/channel-mirrors): OCI mirror of the conda-forge channel

:::info
These organizations exist but they are not in active use anymore:

- [`conda-forge-abandoned`](https://github.com/conda-forge-abandoned)
- [`conda-forge-woodpecker-admins`](https://github.com/conda-forge-woodpecker-admins)

:::

#### Teams

The `conda-forge` Github organization has thousands of teams.
Most of them are associated with a feedstock, but there are a few special ones that are not!

- [`Core`](https://github.com/orgs/conda-forge/teams/Core)
- [`staged-recipes`](https://github.com/orgs/conda-forge/teams/staged-recipes)

> - [`help-c-cpp`](https://github.com/orgs/conda-forge/teams/help-c-cpp)
> - [`help-cdts`](https://github.com/orgs/conda-forge/teams/help-cdts)
> - [`help-go`](https://github.com/orgs/conda-forge/teams/help-go)
> - [`help-java`](https://github.com/orgs/conda-forge/teams/help-java)
> - [`help-julia`](https://github.com/orgs/conda-forge/teams/help-julia)
> - [`help-nodejs`](https://github.com/orgs/conda-forge/teams/help-nodejs)
> - [`help-osx-arm64`](https://github.com/orgs/conda-forge/teams/help-osx-arm64)
> - [`help-perl`](https://github.com/orgs/conda-forge/teams/help-perl)
> - [`help-ppc64le`](https://github.com/orgs/conda-forge/teams/help-ppc64le)
> - [`help-pypy`](https://github.com/orgs/conda-forge/teams/help-pypy)
> - [`help-python`](https://github.com/orgs/conda-forge/teams/help-python)
> - [`help-python-c`](https://github.com/orgs/conda-forge/teams/help-python-c)
> - [`help-r`](https://github.com/orgs/conda-forge/teams/help-r)
> - [`help-ruby`](https://github.com/orgs/conda-forge/teams/help-ruby)

Reviewer 3: It's already out-of-date ;)
```suggestion
- [`help-c-cpp`](https://github.com/orgs/conda-forge/teams/help-c-cpp)
- [`help-cdts`](https://github.com/orgs/conda-forge/teams/help-cdts)
- [`help-go`](https://github.com/orgs/conda-forge/teams/help-go)
- [`help-java`](https://github.com/orgs/conda-forge/teams/help-java)
- [`help-julia`](https://github.com/orgs/conda-forge/teams/help-julia)
- [`help-nodejs`](https://github.com/orgs/conda-forge/teams/help-nodejs)
- [`help-osx-arm64`](https://github.com/orgs/conda-forge/teams/help-osx-arm64)
- [`help-perl`](https://github.com/orgs/conda-forge/teams/help-perl)
- [`help-ppc64le`](https://github.com/orgs/conda-forge/teams/help-ppc64le)
- [`help-pypy`](https://github.com/orgs/conda-forge/teams/help-pypy)
- [`help-python`](https://github.com/orgs/conda-forge/teams/help-python)
- [`help-python-c`](https://github.com/orgs/conda-forge/teams/help-python-c)
- [`help-r`](https://github.com/orgs/conda-forge/teams/help-r)
- [`help-ruby`](https://github.com/orgs/conda-forge/teams/help-ruby)
- [`help-rust`](https://github.com/orgs/conda-forge/teams/help-rust)
```
